io/io/src/TBufferJSON.cxx

```diff
@@ -1889,8 +1889,13 @@ void *TBufferJSON::JsonReadObject(void *obj, const TClass *objClass, TClass **re
          jsonClassVersion = json->at(fTypeVersionTag.Data()).get<int>();
 
       if (objClass && (jsonClass != objClass)) {
-         Error("JsonReadObject", "Class mismatch between provided %s and in JSON %s", objClass->GetName(),
-               jsonClass->GetName());
+         if (obj || (jsonClass->GetBaseClassOffset(objClass) != 0)) {
```

Comment:
This test restrict to the same class or one that has the class as a left most base class.  Is that the intent?
If not, you could still adjust the pointer to support non-left most base classes.

Reply:
Yes, this is intended. I have a use case with `std::unique_ptr<BaseClass>` and want to read objects of derived classes. I am not sure that here complex muliple inheritance need to be supported

Reply:
Supporting the more complex case is (seemingly) straightforward (almost `obj = (...)((char*)obj)+jsonClass->GetBaseClassOffset(objClass)`)

Alternatively the message must be much clearer (i.e. "why" is there a mismatch)

Reply:
This change has some other side effects. I prefer to support here only direct left most inheritance.
I tried to update comment.